man/nemesis-igmp.1
.\" THE NEMESIS PROJECT
.\" Copyright (C) 1999, 2000 Mark Grimes <mark@stateful.net>
.\" Copyright (C) 2001 - 2003 Jeff Nathan <jeff@snort.org>
.\" Copyright (C) 2019 Joachim Nilsson <troglobit@gmail.com>
.\"
.Dd Dec 13, 2019
.Dt nemesis-igmp 1 USM
.Sh NAME
.Nm nemesis-igmp
.Nd IGMP Protocol (The Nemesis Project)
.Sh SYNOPSIS
.Nm
.Op Fl vZ?
.Op Fl c Ar COUNT
.Op Fl d Ar IFNAME
.Op Fl D Ar ADDR
.Op Fl F Ar OPT
.Op Fl g Ar GROUP
.Op Fl H Ar MAC
.Op Fl I Ar ID
.Op Fl i Ar INTERVAL
.Op Fl M Ar MAC
.Op Fl O Ar FILE
.Op Fl p Ar TYPE
.Op Fl P Ar FILE
.Op Fl S Ar ADDR
.Op Fl t Ar TOS
.Op Fl T Ar TTL
.Sh DESCRIPTION
.Nm nemesis
is designed to be a command line-based, portable human IP stack for UNIX-like 
and Windows systems.  The suite is broken down by protocol, and should allow 
for useful scripting of injected packets from simple shell scripts.
.Pp
.Nm
provides an interface to craft and inject IGMP packets allowing the user to
specify any portion of an IGMP packet as well as lower-level IP packet
information.
.Sh GENERAL OPTIONS
.Bl -tag -width Ds
.It Fl c Ar COUNT
Number of packets to send, default: 1.
.It Fl i Ar INTERVAL
Seconds between repeatedly sent packets, only available if
.Fl c
is given.
.It Fl v
Display the injected packet in human readable form.  Use twice to see a
hexdump of the injected packet with printable ASCII characters on the
right.  Use three times for a hexdump without decoded ASCII.
.El
.Sh IGMP Options
.Bl -tag -width Ds
.It Fl r Ar CODE
Specify the IGMP response code, the value for the one-byte field
following the IGMP type field.  Unused in IGMP v1, query response
time in IGMP v2.
.It Fl g Ar GROUP
Specify the IGMP group address within the IGMP header.  It is used for
IGMP v1/v2 join/leave (v1 does not have leave), and group specific IGMP
queries.
.It Fl p Ar TYPE
Specify the IGMP type within the IGMP header.  Example values include:
.Pp
.Bl -tag -width 0x31: -offset indent -compact
.It 0x11:
Query, length determines version
.It 0x12:
Join, v1
.It 0x13:
DVMRP
.It 0x14:
PIM, v1
.It 0x16:
Join, v2
.It 0x17:
Leave, v2
.It 0x1e:
Multicast traceroute, response
.It 0x1f:
Multicast traceroute
.It 0x22:
Membership report, v3 join/leave
.It 0x30:
Multicast router advertisement
.It 0x31:
Multicast router solicitation
.It 0x32:
Multicast router termination
.El
.Pp
Only one type may be specified at a time.
.It Fl P Ar FILE
This will cause
.Nm
to use the specified payload file as the raw payload when injecting IGMP
packets.  For packets injected using the interface (where
.Fl d
is not used), the maximum payload size is 65467 bytes.  For packets
injected using the link layer interface (where
.Fl d
IS used), the maximum payload size is 1432 bytes.  Payloads can also be
read from stdin by specifying
.Fl P-
instead.
.Pp
Windows systems are limited to a maximum payload size of 1432 bytes for
IGMP packets.
.El
.Sh IP OPTIONS
.Bl -tag -width Ds
.It Fl D Ar ADDR
Specify the destination IP address within the IP header.
.It Fl F Ar OPT
Specify the fragmentation options in the IP header:
.Pp
.Bl -tag -width "-F offset" -compact -offset indent
.It Fl FD
don't fragment
.It Fl FM
more fragments
.It Fl FR
reserved flag
.It Fl F Ar offset
.El
.Pp
IP fragmentation options can be specified individually or combined into
a single argument to the
.Fl F
command line switch by separating the options with commas (eg.
.Fl FD,M )
or spaces (eg.
.Fl FM Ar 223 ) .
The IP fragmentation offset is a 13-bit field with valid values from 0
to 8189.  Don't fragment (DF), more fragments (MF) and the reserved flag
(RESERVED or RB) are 1-bit fields.
.Pp
.Sy NOTE:
Under normal conditions, the reserved flag is unset.
.It Fl I Ar ID
Specify the IP ID within the IP header.
.It Fl O Ar FILE
This will cause
.Nm
to use the specified IP options file as the options when building the IP
header for the injected packet.  IP options can be up to 40 bytes in
length.  The IP options file must be created manually based upon the
desired options.  IP options can also be read from stdin by specifying
.Fl O-
instead.
.It Fl S Ar ADDR
Specify the source IP address within the IP header.
.It Fl t Ar TOS
Specify the IP type of service (TOS) within the IP header.  Valid type
of service values:
.Pp
.Bl -tag -width 24 -offset indent -compact
.It 2
Minimize monetary cost
.It 4
Maximize reliability
.It 8
Maximize throughput
.It 24
Minimize delay
.El
.Pp
.Sy NOTE:
Under normal conditions, only one type of service is set within a
packet.  To specify multiple types, specify the sum of the desired
values as the type of service.
.It Fl T Ar TTL
Specify the IP time-to-live (TTL) in the IP header.
.El
.Sh DATA LINK OPTIONS
.Bl -tag -width Ds
.It Fl d Ar IFNAME
Specify the name (for UNIX-like systems) or the number (for Windows
systems) of the
.Ar IFNAME
to use (eg. fxp0, eth0, hme0, 1).
.It Fl H Ar MAC
Specify the source
.Ar MAC
address,
.Ar ( XX:XX:XX:XX:XX:XX ) .
.It Fl M Ar MAC
Specify the destination
.Ar MAC
address,
.Ar ( XX:XX:XX:XX:XX:XX ) .
.It Fl Z
Lists the available network interfaces by number for use in link-layer
injection.
.Pp
.Sy NOTE:
This feature is only relevant to Windows systems.
.El
.Sh DIAGNOSTICS
.Nm
returns 0 on a successful exit, 1 if it exits on an error.
.Sh SEE ALSO
.Xr nemesis-arp 1 ,
.Xr nemesis-dhcp 1 ,
.Xr nemesis-dns 1 ,
.Xr nemesis-ethernet 1 ,
.Xr nemesis-icmp 1 ,
.Xr nemesis-ip 1 ,
.Xr nemesis-ospf 1 ,
.Xr nemesis-rip 1 ,
.Xr nemesis-tcp 1 ,
.Xr nemesis-udp 1 .
.Sh AUTHORS
.An Mark Grimes Aq Mt mark@stateful.net ,
.An Jeff Nathan Aq Mt jeff@snort.org
and
.An Joachim Nilsson Aq Mt troglobit@gmail.com
.Sh BUGS
Please report at
.Lk https://github.com/libnet/nemesis/issues


src/nemesis-igmp.c
/*
 * THE NEMESIS PROJECT
 * Copyright (C) 1999, 2000, 2001 Mark Grimes <mark@stateful.net>
 * Copyright (C) 2001 - 2003 Jeff Nathan <jeff@snort.org>
 *
 * nemesis-igmp.c (IGMP Packet Injector)
 */

#include "nemesis-igmp.h"
#include "nemesis.h"
#if defined(WIN32)
#include <pcap.h>
#endif

static ETHERhdr etherhdr;
static IPhdr    iphdr;
static IGMPhdr  igmphdr;
static struct file pd, ipod;
static int      got_group, got_type, got_code;
static char    *payloadfile   = NULL; /* payload file name */
static char    *ipoptionsfile = NULL; /* IP options file name */
static char    *device        = NULL; /* Ethernet device */

#if defined(WIN32)
static char *ifacetmp = NULL;
#endif

static void igmp_cmdline(int, char **);
static int  igmp_exit(int);
static void igmp_initdata(void);
static void igmp_usage(char *);
static void igmp_validatedata(libnet_t *l);
static void igmp_verbose(void);

void nemesis_igmp(int argc, char **argv)
{
	const char *module = "IGMP Packet Injection";
	libnet_t *l;

	nemesis_maketitle(title, module, version);

	if (argc > 1 && !strncmp(argv[1], "help", 4))
		igmp_usage(argv[0]);

	if (nemesis_seedrand() < 0)
		fprintf(stderr, "ERROR: Unable to seed random number generator.\n");

	igmp_initdata();
	igmp_cmdline(argc, argv);

	l = libnet_init(got_link ? LIBNET_LINK_ADV : LIBNET_RAW4, device, errbuf);
	if (!l)
		igmp_exit(1);

	igmp_validatedata(l);
	igmp_verbose();

	if (got_payload) {
#if defined(WIN32)
		if (builddatafromfile(IGMP_LINKBUFFSIZE, &pd, payloadfile, PAYLOADMODE) < 0)
#else
		if (builddatafromfile(((got_link == 1) ? IGMP_LINKBUFFSIZE : IGMP_RAWBUFFSIZE), &pd, payloadfile, PAYLOADMODE) < 0)
#endif
			igmp_exit(1);
	}

	if (got_ipoptions) {
		if (builddatafromfile(OPTIONSBUFFSIZE, &ipod, ipoptionsfile, OPTIONSMODE) < 0)
			igmp_exit(1);
	}

	if (buildigmp(&etherhdr, &iphdr, &igmphdr, &pd, &ipod, l) < 0) {
		puts("\nIGMP Injection Failure");
		igmp_exit(1);
	}

	igmp_exit(0);
}

static void igmp_initdata(void)
{
	/* defaults */
	etherhdr.ether_type = ETHERTYPE_IP;    /* Ethernet type IP */
	memset(etherhdr.ether_shost, 0, 6);    /* Ethernet source address */
	memset(etherhdr.ether_dhost, 0xff, 6); /* Ethernet destination address */

	iphdr.ip_src.s_addr = libnet_get_prand(PRu32); /* IP source address */
	memset(&iphdr.ip_dst.s_addr, 0, 4);            /* IP destination address */
	iphdr.ip_tos = 0;                              /* IP type of service */
	iphdr.ip_id  = libnet_get_prand(PRu16);        /* IP ID */
	iphdr.ip_p   = IPPROTO_IGMP;                   /* IP protocol IGMP */
	iphdr.ip_off = 0;                              /* IP fragmentation offset */
	iphdr.ip_ttl = 1;                              /* IP TTL - set to 1 purposely */

	igmphdr.igmp_type         = 0; /* IGMP type */
	igmphdr.igmp_code         = 0; /* IGMP code */
	igmphdr.igmp_group.s_addr = 0; /* IGMP group IP address */

	pd.file_buf   = NULL;
	pd.file_len   = 0;
	ipod.file_buf = NULL;
	ipod.file_len = 0;
}

static void igmp_validatedata(libnet_t *l)
{
	/*
	 * if a device was specified and the user has not specified a
	 * source hardware address, try to determine the source address
	 * automatically
	 */
	if (got_link) {
		if ((nemesis_check_link(&etherhdr, l)) < 0) {
			fprintf(stderr, "ERROR: Cannot retrieve hardware address of %s.\n", device);
			igmp_exit(1);
		}
	}

	/* Attempt to send valid packets if the user hasn't decided to craft an anomolous packet */
	if (!got_type)
		igmphdr.igmp_type = IGMP_V1_MEMBERSHIP_REPORT;
	if (!got_code)
		igmphdr.igmp_code = 0;
	if (!got_group) {
		inet_aton("224.0.0.1", &igmphdr.igmp_group);
		inet_aton("224.0.0.1", &iphdr.ip_dst);
	}
}

static void igmp_usage(char *arg)
{
	nemesis_printtitle(title);

	printf("Usage:\n"
	       "  %s [-v (verbose)] [options]\n"
	       "\n", arg);
	printf("General Options:\n"
	       "  -c <COUNT>   Send count number of packets\n"
	       "  -i <WAIT>    Interval to wait between packets\n"
	       "\n");
	printf("IGMP options:\n"
	       "  -p <TYPE>    IGMP protocol type:\n"
	       "                    0x11:  Query, length determines version\n"
	       "                    0x12:  Join, v1\n"
	       "                    0x13:  DVMRP\n"
	       "                    0x14:  PIM, v1\n"
	       "                    0x16:  Join, v2\n"
	       "                    0x17:  Leave, v2\n"
	       "                    0x1e:  Multicast traceroute, response\n"
	       "                    0x1f:  Multicast traceroute\n"
	       "                    0x22:  Membership report, v3 join/leave\n"
	       "                    0x30:  Multicast router advertisement\n"
	       "                    0x31:  Multicast router solicitation\n"
	       "                    0x32:  Multicast router termination\n"
	       "  -r <CODE>    Max resp. code. v1: unused, v2: query response time\n"
	       "  -g <GROUP>   Multicast group for join/leave, or group spec. query\n"
	       "  -P <FILE>    Raw IGMP payload file\n"
	       "\n");
	printf("IP options:\n"
	       "  -S <ADDR>    Source IP address\n"
	       "  -D <ADDR>    Destination IP address\n"
	       "  -I <ID>      IP ID\n"
	       "  -T <TTL>     IP TTL\n"
	       "  -t <TOS>     IP TOS\n"
	       "  -F <OPT>     IP fragmentation options: -F[D],[M],[R],[offset]\n"
	       "  -O <FILE>    Raw IP options file\n"
	       "\n");
	printf("Data Link Options:\n"
#if defined(WIN32)
	       "  -d <IFNUM>   Network interface number\n"
#else
	       "  -d <IFNAME>  Network interface name\n"
#endif
	       "  -H <MAC>     Source MAC address\n"
	       "  -M <MAC>     Destination MAC address\n");
#if defined(WIN32)
	printf("  -Z           List available network interfaces by number\n");
#endif
	putchar('\n');
	igmp_exit(1);
}

static void igmp_cmdline(int argc, char **argv)
{
	int          opt, i;
	uint32_t     addr_tmp[6];
	char        *igmp_options;

#if defined(WIN32)
	igmp_options = "c:d:D:F:g:H:i:I:M:O:p:P:r:S:t:T:vZ?";
#else
	igmp_options = "c:d:D:F:g:H:i:I:M:O:p:P:r:S:t:T:v?";
#endif
	while ((opt = getopt(argc, argv, igmp_options)) != -1) {
		switch (opt) {
		case 'c':
			count = atoi(optarg);
			break;

		case 'i':
			interval = xgetusec(optarg);
			break;

		case 'd': /* Ethernet device */
#if defined(WIN32)
			if (nemesis_getdev(atoi(optarg), &device) < 0) {
				fprintf(stderr, "ERROR: Unable to lookup device: '%d'.\n", atoi(optarg));
				igmp_exit(1);
			}
#else
			if (strlen(optarg) < 256) {
				if (device)
					free(device);
				device = strdup(optarg);
				got_link = 1;
			} else {
				fprintf(stderr, "ERROR: device %s > 256 characters.\n", optarg);
				igmp_exit(1);
			}
#endif
			break;

		case 'D': /* destination IP address */
			if ((nemesis_name_resolve(optarg, &iphdr.ip_dst.s_addr)) < 0) {
				fprintf(stderr, "ERROR: Invalid destination IP address: \"%s\".\n", optarg);
				igmp_exit(1);
			}
			break;

		case 'F': /* IP fragmentation options */
			if (parsefragoptions(&iphdr, optarg) < 0)
				igmp_exit(1);
			break;

		case 'g': /* IGMP group address */
			if ((nemesis_name_resolve(optarg, &igmphdr.igmp_group.s_addr)) < 0) {
				fprintf(stderr, "ERROR: Invalid IGMP group address: \"%s\".\n", optarg);
				igmp_exit(1);
			}
			got_group = 1;
			break;

		case 'H': /* Ethernet source address */
			memset(addr_tmp, 0, sizeof(addr_tmp));
			sscanf(optarg, "%02X:%02X:%02X:%02X:%02X:%02X", &addr_tmp[0],
			       &addr_tmp[1], &addr_tmp[2], &addr_tmp[3], &addr_tmp[4], &addr_tmp[5]);
			for (i = 0; i < 6; i++)
				etherhdr.ether_shost[i] = addr_tmp[i];
			break;

		case 'I': /* IP ID */
			iphdr.ip_id = xgetint16(optarg);
			break;

		case 'M': /* Ethernet destination address */
			memset(addr_tmp, 0, sizeof(addr_tmp));
			sscanf(optarg, "%02X:%02X:%02X:%02X:%02X:%02X", &addr_tmp[0],
			       &addr_tmp[1], &addr_tmp[2], &addr_tmp[3], &addr_tmp[4], &addr_tmp[5]);
			for (i = 0; i < 6; i++)
				etherhdr.ether_dhost[i] = addr_tmp[i];
			break;

		case 'O': /* IP options file */
			if (strlen(optarg) < 256) {
				if (ipoptionsfile)
					free(ipoptionsfile);
				ipoptionsfile = strdup(optarg);
				got_ipoptions = 1;
			} else {
				fprintf(stderr, "ERROR: IP options file %s > 256 characters.\n", optarg);
				igmp_exit(1);
			}
			break;

		case 'p': /* IGMP type */
			igmphdr.igmp_type = xgetint8(optarg);
			got_type          = 1;
			break;

		case 'P': /* payload file */
			if (strlen(optarg) < 256) {
				if (payloadfile)
					free(payloadfile);
				payloadfile = strdup(optarg);
				got_payload = 1;
			} else {
				fprintf(stderr, "ERROR: payload file %s > 256 characters.\n", optarg);
				igmp_exit(1);
			}
			break;

		case 'r': /* IGMP max resp. code */
			igmphdr.igmp_code = xgetint8(optarg);
			got_code          = 1;
			break;

		case 'S': /* source IP address */
			if ((nemesis_name_resolve(optarg, &iphdr.ip_src.s_addr)) < 0) {
				fprintf(stderr, "ERROR: Invalid source IP address: \"%s\".\n", optarg);
				igmp_exit(1);
			}
			break;

		case 't': /* IP type of service */
			iphdr.ip_tos = xgetint8(optarg);
			break;

		case 'T': /* IP time to live */
			iphdr.ip_ttl = xgetint8(optarg);
			break;

		case 'v':
			verbose++;
			if (verbose == 1)
				nemesis_printtitle(title);
			break;
#if defined(WIN32)
		case 'Z':
			if ((ifacetmp = pcap_lookupdev(errbuf)) == NULL)
				perror(errbuf);

			PrintDeviceList(ifacetmp);
			igmp_exit(1);
#endif
		case '?': /* FALLTHROUTH */
		default:
			igmp_usage(argv[0]);
			break;
		}
	}
	argc -= optind;
	argv += optind;
}

static int igmp_exit(int code)
{
	if (got_payload)
		free(pd.file_buf);

	if (got_ipoptions)
		;
	free(ipod.file_buf);

	if (device != NULL)
		free(device);

	if (ipoptionsfile != NULL)
		free(ipoptionsfile);

	if (payloadfile != NULL)
		free(payloadfile);

#if defined(WIN32)
	if (ifacetmp != NULL)
		free(ifacetmp);
#endif

	exit(code);
}

static void igmp_verbose(void)
{
	if (verbose) {
		if (got_link)
			nemesis_printeth(&etherhdr);

		nemesis_printip(&iphdr);
		printf("         [IGMP Type] %hhu\n", igmphdr.igmp_type);
		printf("         [IGMP Code] %hhu\n", igmphdr.igmp_code);
		printf("[IGMP group address] %s\n", inet_ntoa(igmphdr.igmp_group));
	}
}
